third/sendmail/src/sendmail.8
.\" Dummy comment to avoid confusing man
.so /usr/athena/share/tmac/tmac.doc
.\" Copyright (c) 1998 Sendmail, Inc.  All rights reserved.
.\" Copyright (c) 1983, 1997 Eric P. Allman.  All rights reserved.
.\" Copyright (c) 1988, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" By using this file, you agree to the terms and conditions set
.\" forth in the LICENSE file which can be found at the top level of
.\" the sendmail distribution.
.\"
.\"
.\"     @(#)sendmail.8	8.20 (Berkeley) 8/2/1998
.\"
.Dd August 2, 1998
.Dt SENDMAIL 8
.Os BSD 4
.Sh NAME
.Nm sendmail
.Nd an electronic mail transport agent
.Sh SYNOPSIS
.Nm sendmail
.Op Ar flags
.Op Ar address ...
.Nm newaliases
.Nm mailq
.Op Fl v
.Sh DESCRIPTION
.Nm Sendmail
sends a message to one or more
.Em recipients ,
routing the message over whatever networks
are necessary.
.Nm Sendmail
does internetwork forwarding as necessary
to deliver the message to the correct place.
.Pp
.Nm Sendmail
is not intended as a user interface routine;
other programs provide user-friendly
front ends;
.Nm sendmail
is used only to deliver pre-formatted messages.
.Pp
With no flags,
.Nm sendmail
reads its standard input
up to an end-of-file
or a line consisting only of a single dot
and sends a copy of the message found there
to all of the addresses listed.
It determines the network(s) to use
based on the syntax and contents of the addresses.
.Pp
Local addresses are looked up in a file
and aliased appropriately.
Aliasing can be prevented by preceding the address
with a backslash.
Normally the sender is not included in any alias
expansions, e.g.,
if `john' sends to `group',
and `group' includes `john' in the expansion,
then the letter will not be delivered to `john'.
.Ss Parameters
.Bl -tag -width Fl
.It Fl B Ns Ar type
Set the body type to
.Ar type .
Current legal values
.Li 7BIT
or
.Li 8BITMIME .
.It Fl ba
Go into
.Tn ARPANET
mode.
All input lines must end with a CR-LF,
and all messages will be generated with a CR-LF at the end.
Also,
the ``From:'' and ``Sender:''
fields are examined for the name of the sender.
.It Fl bd
Run as a daemon.  This requires Berkeley
.Tn IPC .
.Nm Sendmail
will fork and run in background
listening on socket 25 for incoming
.Tn SMTP
connections.
This is normally run from
.Pa /etc/rc .
.It Fl bD
Same as
.Fl bd
except runs in foreground.
.It Fl bh
Print the persistent host status database.
.It Fl bH
Purge the persistent host status database.
.It Fl bi
Initialize the alias database.
.It Fl bm
Deliver mail in the usual way (default).
.It Fl bp
Print a listing of the queue.
.It Fl bs
Use the
.Tn SMTP
protocol as described in
.Tn RFC821
on standard input and output.
This flag implies all the operations of the
.Fl ba
flag that are compatible with
.Tn SMTP .
.It Fl bt
Run in address test mode.
This mode reads addresses and shows the steps in parsing;
it is used for debugging configuration tables.
.It Fl bv
Verify names only \- do not try to collect or deliver a message.
Verify mode is normally used for validating
users or mailing lists.
.It Fl C Ns Ar file
Use alternate configuration file.
.Nm Sendmail
refuses to run as root if an alternate configuration file is specified.
.It Fl d Ns Ar X
Set debugging value to
.Ar X .
.ne 1i
.It Fl F Ns Ar fullname
Set the full name of the sender.
.It Fl f Ns Ar name
Sets the name of the ``from'' person
(i.e., the sender of the mail).
.Fl f
can only be used
by ``trusted'' users
(normally
.Em root ,
.Em daemon ,
and
.Em network )
or if the person you are trying to become
is the same as the person you are.
.It Fl h Ns Ar N
Set the hop count to
.Ar N .
The hop count is incremented every time the mail is
processed.
When it reaches a limit,
the mail is returned with an error message,
the victim of an aliasing loop.
If not specified,
``Received:'' lines in the message are counted.
.It Fl i
Ignore dots alone on lines by themselves in incoming messages.
This should be set if you are reading data from a file.
.It Fl N Ar dsn
Set delivery status notification conditions to
.Ar dsn,
which can be
.Ql never
for no notifications
or a comma separated list of the values
.Ql failure
to be notified if delivery failed,
.Ql delay
to be notified if delivery is delayed, and
.Ql success
to be notified when the message is successfully delivered.
.It Fl n
Don't do aliasing.
.It Fl O Ar option Ns = Ns Em value
Set option
.Ar option
to the specified
.Em value .
This form uses long names.
See below for more details.
.It Fl o Ns Ar x Em value
Set option
.Ar x
to the specified
.Em value .
This form uses single character names only.
The short names are not described in this manual page;
see the
.%T "Sendmail Installation and Operation Guide"
for details.
.It Fl p Ns Ar protocol
Set the name of the protocol used to receive the message.
This can be a simple protocol name such as ``UUCP''
or a protocol and hostname, such as ``UUCP:ucbvax''.
.It Fl q Ns Bq Ar time
Processed saved messages in the queue at given intervals.
If
.Ar time
is omitted,
process the queue once.
.Xr Time
is given as a tagged number,
with
.Ql s
being seconds,
.Ql m
being minutes,
.Ql h
being hours,
.Ql d
being days,
and
.Ql w
being weeks.
For example,
.Ql \-q1h30m
or
.Ql \-q90m
would both set the timeout to one hour thirty minutes.
If
.Ar time
is specified,
.Nm sendmail
will run in background.
This option can be used safely with
.Fl bd .
.It Fl qI Ns Ar substr
Limit processed jobs to those containing
.Ar substr
as a substring of the queue id.
.It Fl qR Ns Ar substr
Limit processed jobs to those containing
.Ar substr
as a substring of one of the recipients.
.It Fl qS Ns Ar substr
Limit processed jobs to those containing
.Ar substr
as a substring of the sender.
.It Fl R Ar return
Set the amount of the message to be returned
if the message bounces.
The
.Ar return
parameter can be
.Ql full
to return the entire message or
.Ql hdrs
to return only the headers.
.It Fl r Ns Ar name
An alternate and obsolete form of the
.Fl f
flag.
.It Fl t
Read message for recipients.
To:, Cc:, and Bcc: lines will be scanned for recipient addresses.
The Bcc: line will be deleted before transmission.
.It Fl U
Initial (user) submission.
This should
.Em always
be set when called from a user agent such as
.Nm Mail
or
.Nm exmh
and
.Em never
be set when called by a network delivery agent such as
.Nm rmail .
.It Fl V Ar envid
Set the original envelope id.
This is propagated across SMTP to servers that support DSNs
and is returned in DSN-compliant error messages.
.It Fl v
Go into verbose mode.
Alias expansions will be announced, etc.
.It Fl X Ar logfile
Log all traffic in and out of mailers in the indicated log file.
This should only be used as a last resort
for debugging mailer bugs.
It will log a lot of data very quickly.
.It Fl -
Stop processing command flags and use the rest of the arguments
as addresses.
.El
.Ss Options
There are also a number of processing options that may be set.
Normally these will only be used by a system administrator.
Options may be set either on the command line
using the
.Fl o
flag (for short names),
the
.Fl O
flag (for long names),
or in the configuration file.
This is a partial list limited to those options that are likely to be useful
on the command line
and only shows the long names;
for a complete list (and details), consult the
.%T "Sendmail Installation and Operation Guide" .
The options are:
.Bl -tag -width Fl
.It Li AliasFile= Ns Ar file
Use alternate alias file.
.It Li HoldExpensive
On mailers that are considered ``expensive'' to connect to,
don't initiate immediate connection.
This requires queueing.
.It Li CheckpointInterval= Ns Ar N
Checkpoint the queue file after every
.Ar N
successful deliveries (default 10).
This avoids excessive duplicate deliveries
when sending to long mailing lists
interrupted by system crashes.
.ne 1i
.It Li DeliveryMode= Ns Ar x
Set the delivery mode to
.Ar x .
Delivery modes are
.Ql i
for interactive (synchronous) delivery,
.Ql b
for background (asynchronous) delivery,
.Ql q
for queue only \- i.e.,
actual delivery is done the next time the queue is run, and
.Ql d
for deferred \- the same as
.Ql q
except that database lookups (notably DNS and NIS lookups) are avoided.
.It Li ErrorMode= Ns Ar x
Set error processing to mode
.Ar x .
Valid modes are
.Ql m
to mail back the error message,
.Ql w
to ``write'' back the error message
(or mail it back if the sender is not logged in),
.Ql p
to print the errors on the terminal
(default),
.Ql q
to throw away error messages
(only exit status is returned),
and
.Ql e
to do special processing for the BerkNet.
If the text of the message is not mailed back
by
modes
.Ql m
or
.Ql w
and if the sender is local to this machine,
a copy of the message is appended to the file
.Pa dead.letter
in the sender's home directory.
.It Li SaveFromLine
Save
.Tn UNIX Ns \-style
From lines at the front of messages.
.It Li MaxHopCount= Ar N
The maximum number of times a message is allowed to ``hop''
before we decide it is in a loop.
.It Li IgnoreDots
Do not take dots on a line by themselves
as a message terminator.
.It Li SendMimeErrors
Send error messages in MIME format.
If not set, the DSN (Delivery Status Notification) SMTP extension
is disabled.
.It Li ConnectionCacheTimeout= Ns Ar timeout
Set connection cache timeout.
.It Li ConnectionCacheSize= Ns Ar N
Set connection cache size.
.It Li LogLevel= Ns Ar n
The log level.
.It Li MeToo
Send to ``me'' (the sender) also if I am in an alias expansion.
.It Li CheckAliases
Validate the right hand side of aliases during a
.Xr newaliases 1
command.
.It Li OldStyleHeaders
If set, this message may have
old style headers.
If not set,
this message is guaranteed to have new style headers
(i.e., commas instead of spaces between addresses).
If set, an adaptive algorithm is used that will correctly
determine the header format in most cases.
.It Li QueueDirectory= Ns Ar queuedir
Select the directory in which to queue messages.
.It Li StatusFile= Ns Ar file
Save statistics in the named file.
.It Li Timeout.queuereturn= Ns Ar time
Set the timeout on undelivered messages in the queue to the specified time.
After delivery has failed
(e.g., because of a host being down)
for this amount of time,
failed messages will be returned to the sender.
The default is five days.
.It Li UserDatabaseSpec= Ns Ar userdatabase
If set, a user database is consulted to get forwarding information.
You can consider this an adjunct to the aliasing mechanism,
except that the database is intended to be distributed;
aliases are local to a particular host.
This may not be available if your sendmail does not have the
.Dv USERDB
option compiled in.
.It Li ForkEachJob
Fork each job during queue runs.
May be convenient on memory-poor machines.
.It Li SevenBitInput
Strip incoming messages to seven bits.
.It Li EightBitMode= Ns Ar mode
Set the handling of eight bit input to seven bit destinations to
.Ar mode :
.Li m
(mimefy) will convert to seven-bit MIME format,
.Li p
(pass) will pass it as eight bits (but violates protocols),
and
.Li s
(strict) will bounce the message.
.It Li MinQueueAge= Ns Ar timeout
Sets how long a job must ferment in the queue between attempts to send it.
.It Li DefaultCharSet= Ns Ar charset
Sets the default character set used to label 8-bit data
that is not otherwise labelled.
.It Li DialDelay= Ns Ar sleeptime
If opening a connection fails,
sleep for
.Ar sleeptime
seconds and try again.
Useful on dial-on-demand sites.
.It Li NoRecipientAction= Ns Ar action
Set the behaviour when there are no recipient headers (To:, Cc: or Bcc:)
in the message to
.Ar action :
.Li none
leaves the message unchanged,
.Li add-to
adds a To: header with the envelope recipients,
.Li add-apparently-to
adds an Apparently-To: header with the envelope recipients,
.Li add-bcc
adds an empty Bcc: header, and
.Li add-to-undisclosed
adds a header reading
.Ql "To: undisclosed-recipients:;" .
.It Li MaxDaemonChildren= Ns Ar N
Sets the maximum number of children that an incoming SMTP daemon
will allow to spawn at any time to
.Ar N .
.It Li ConnectionRateThrottle= Ns Ar N
Sets the maximum number of connections per second to the SMTP port to
.Ar N .
.El
.Pp
In aliases,
the first character of a name may be
a vertical bar to cause interpretation of
the rest of the name as a command
to pipe the mail to.
It may be necessary to quote the name
to keep
.Nm sendmail
from suppressing the blanks from between arguments.
For example, a common alias is:
.Pp
.Bd -literal -offset indent -compact
msgs: "|/usr/bin/msgs -s"
.Ed
.Pp
Aliases may also have the syntax
.Dq :include: Ns Ar filename
to ask
.Xr sendmail
to read the named file for a list of recipients.
For example, an alias such as:
.Pp
.Bd -literal -offset indent -compact
poets: ":include:/usr/local/lib/poets.list"
.Ed
.Pp
would read
.Pa /usr/local/lib/poets.list
for the list of addresses making up the group.
.Pp
.Nm Sendmail
returns an exit status
describing what it did.
The codes are defined in
.Aq Pa sysexits.h :
.Bl -tag -width EX_UNAVAILABLE -compact -offset indent
.It Dv EX_OK
Successful completion on all addresses.
.It Dv EX_NOUSER
User name not recognized.
.It Dv EX_UNAVAILABLE
Catchall meaning necessary resources
were not available.
.It Dv EX_SYNTAX
Syntax error in address.
.It Dv EX_SOFTWARE
Internal software error,
including bad arguments.
.It Dv EX_OSERR
Temporary operating system error,
such as
.Dq cannot fork .
.It Dv EX_NOHOST
Host name not recognized.
.It Dv EX_TEMPFAIL
Message could not be sent immediately,
but was queued.
.El
.Pp
If invoked as
.Nm newaliases ,
.Nm sendmail
will rebuild the alias database.
If invoked as
.Nm mailq ,
.Nm sendmail
will print the contents of the mail queue.
.Sh FILES
Except for the file
.Pa /etc/sendmail.cf
itself and the daemon process ID file,
the following pathnames are all specified in
.Pa /etc/sendmail.cf.
Thus,
these values are only approximations.
.Pp
.Bl -tag -width /usr/lib/sendmail.fc -compact
.It Pa /etc/aliases
raw data for alias names
.It Pa /etc/aliases.db
data base of alias names
.It Pa /etc/sendmail.cf
configuration file
.It Pa /etc/sendmail.hf
help file
.It Pa /var/log/sendmail.st
collected statistics
.It Pa /var/spool/mqueue/*
temp files
.El
.Sh SEE ALSO
.Xr binmail 1 ,
.Xr mail 1 ,
.Xr rmail 1 ,
.Xr syslog 3 ,
.Xr aliases 5 ,
.Xr mailaddr 7 ,
.Xr rc 8 ;
.Pp
DARPA
Internet Request For Comments
.%T RFC819 ,
.%T RFC821 ,
.%T RFC822 .
.Rs
.%T "Sendmail \- An Internetwork Mail Router"
.%V SMM
.%N \&No. 9
.Re
.Rs
.%T "Sendmail Installation and Operation Guide"
.%V SMM
.%N \&No. 8
.Re
.Sh HISTORY
The
.Nm
command appeared in
.Bx 4.2 .


doc/news
This file is intended to keep track of user-visible changes to the
release, for the construction of the Athena release notes.

Changes from release 8.0 to release 8.1:

	* Athena version numbers are now dotted triplets, e.g. 8.1.0
	  or 8.1.12.  Paths which include all or part of the Athena
	  version number now take the form "pathname-version",
	  e.g. "srvd-8.1" or "dash.menus-8.1"

	* In release 8.0, running "xmkmf" with the default path would
	  generate a Makefile which looked for its include files and
	  libraries in /mit/x11, which could be one of several
	  lockers.  Now it is preferrable to use the vendor X
	  libraries, so this bit of glue is going away.  There will be
	  no /usr/athena/bin/xmkmf any more, nor any symlinks in
	  /usr/athena pointing to the x11 locker.

	  Users can still do something like "add -f x11r5" and then
	  "xmkmf", and things should work, if the x11r5 locker is
	  built for their platform.

	* Athena bitmaps have moved to /usr/athena/include/X11/bitmaps.
	  Previously, they were in /usr/athena/lib/X11/bitmaps before
	  because of the /mit/x11 glue.

	* The symlink /bin/csh -> /bin/athena/tcsh has gone away.  The
	  update to 8.0 will fix /bin/csh references in /etc/passwd
	  and /etc/passwd.local to refer to /bin/athena/tcsh instead.

	* emacs18 has gone away.

	* AFS includes and libraries are installed in
	  /usr/afsws/include and /usr/afsws/lib.

	* wcl includes and libraries are installed in /usr/athena.

	* The basic krb5 headers and libraries are installed in
	  /usr/athena.

	* emacs has been upgraded from 19.30 to 19.34.  There should
	  be no significant user-visible changes.

	* RCS has been upgraded from version 4 to version 5.7.

	* x3270 has been upgraded from version 1.2 to version 3.1.1.3.

	* sysinfo is now available under Irix.

	* /bin/athena/tcsh now has Hesiod support enabled, which means
	  that "~lockername" will work in a pathname in most cases.
	  (It won't work for users with local passwd information who
	  don't have their lockers attached.)

	* For improved ease of debugging, many binaries are installed
	  with symbol tables, and compiled without optimization.
	  Therefore, the contents of /bin/athena and /usr/athena/bin
	  may be somewhat larger, and the binaries may consume more
	  CPU cycles, although it's unlikely that anyone will notice
	  the speed difference.


third/mh/uip/sendmail.c
/* sendmail.c - */
#ifndef	lint
static char Id[] = "$Id: sendmail.c,v 1.1.1.1 1996-10-07 07:14:21 ghudson Exp $";
#endif
/*
 **  A Sendmail fake.
 *
 * Contributed by Scott Erickson <erickson@ics.uci.edu>
 */
/* Include files glommed from post.c */

#include "../h/mh.h"
#include "../h/addrsbr.h"
#include "../h/aliasbr.h"
#include "../h/dropsbr.h"
#include "../zotnet/tws.h"
#ifndef MMDFMTS
#include <ctype.h>
#include <errno.h>
#include <setjmp.h>
#include <stdio.h>
#include <sys/types.h>
#else   MMDFMTS
#include "../mts/mmdf/util.h"
#include "../mts/mmdf/mmdf.h"
#endif  MMDFMTS
#include "../zotnet/mts.h"
#ifdef  MHMTS
#ifndef V7
#include <sys/ioctl.h>
#endif  not V7
#include <sys/stat.h>
#endif  MHMTS
#ifdef  SENDMTS
#include "../mts/sendmail/smail.h"
#undef  MF
#endif  SENDMTS
#include <signal.h>
#ifdef LOCALE
#include	<locale.h>
#endif

char	*SMTPSRVR = "smtpsrvr";

char	msgfname[50];	/* name of message file */
char	*FullName;	/* sender's full name */
char	*from;		/* sender's mail address */
int	verbose;
int	verify;
int	extract;
int	dodist;
int	rewritefrom;
int	status;		/* return value from procedures */
static int childid;	/* id from smtp child process */
TYPESIG	die();
long	lclock = 0L;	/* the time we started (more or less) */


FILE *fp;		/* file pointer for message file */
extern FILE *tmpfile();

static struct swit switches[] = {
#define ARPASW    0
    "ba", -2,
#define DAEMONSW  1
    "bd", -2,
#define INITALSW  2
    "bi", -2,
#define DELIVSW   3
    "bm", -2,
#define QSUMSW    4
    "bp", -2,
#define SMTPSW    5
    "bs", -2,
#define ADRTSTSW  6
    "bt", -2,
#define ADRVRFSW  7
    "bv", -2,
#define CFGFRZSW  8
    "bz", -2,
#define ALTCFGSW  9
    "C", -1,
#define DBGVALSW 10
    "d", -1,
#define FULLSW   11
    "F", -1,
#define FROMSW   12
    "f", -1,
#define HOPCNTSW 13
    "h", -1,
#define MSGIDSW  14
    "M", -1,
#define NOALISW  15
    "n", -1,
#define QTIMESW  16
    "q", -1,
#define OBSFRMSW 17
    "r", -1,
#define EXTHDRSW 18
    "t", -1,
#define VERBSW	 19
    "v", -1,
#define ALTALISW 20
    "oA", -2,
#define NOCONSW   21
    "oc", -2,
#define DLVMODSW 22
    "od", -2,
#define NEWALISW 23
    "oD", -2,
#define ERRMODSW 24
    "oe", -2,
#define TMPMODSW 25
    "oF", -2,
#define UFROMSW  26
    "of", -2,
#define GIDSW    27
    "og", -2,
#define HLPFILSW 28
    "oH", -2,
#define NODOTSW  29
    "oi", -2,
#define LOGLEVSW 30
    "oL", -2,
#define MEOKSW   31
    "om", -2,
#define OLDHDRSW 32
    "oo", -2,
#define QDIRSW   33
    "oQ", -2,
#define RTMOUTSW 34
    "or", -2,
#define SFILESW  35
    "oS", -2,
#define QMSGSW   36
    "os", -2,
#define MTMOUTSW 37
    "oT", -2,
#define TZSW     38
    "ot", -2,
#define UIDSW    39
    "ou", -2,
    
    NULL, 0
    };

#if !defined(POSIX) && !defined(_POSIX_SOURCE)
extern char *mktemp();
#endif

static void removemsg();
static int  isheader(), sendfile();

/*ARGSUSED*/
main (argc, argv)
int    argc;
char **argv;
{
    register char *cp;
    char **argp = argv + 1;
    
#ifdef LOCALE
	setlocale(LC_ALL, "");
#endif
    invo_name = r1bindex (argv[0], '/');
    mts_init(argv[0]);
    
    if (signal(SIGINT, SIG_IGN) != SIG_IGN)
	(void) signal(SIGINT, die);
    if (signal(SIGHUP, SIG_IGN) != SIG_IGN)
	(void) signal(SIGHUP, die);
    (void) signal(SIGTERM, die);
    (void) signal(SIGPIPE, die);
    
    FullName = getfullname();
    from = adrsprintf(NULLCP,NULLCP);
    (void) time (&lclock);
    
    while ( (cp = *argp) &&  *cp == '-' ) {
	argp++;
	switch (smatch ( ++cp, switches )) {
	case ARPASW:	/* smtp on stdin */
	case SMTPSW:	/* smtp on stdin */
	    smtp();
	    exit(98);	/* should never happen */
	    
	case DELIVSW:	/* just send mail */
	    continue;
	    
	case ADRVRFSW:	/* verify mode */
	    verify = 1;
	    continue;
	    
	case FROMSW:	/* from address */
	case OBSFRMSW:	/* obsolete -f flag */
	    if (*(++cp) == '\0' &&
		(!(cp = *argp++) || *cp == '-'))
		adios (NULLCP, "missing argument to %s", argp[-2]);
	    /* At this point, cp points to the from name */
	    if (rewritefrom) {
		adios (NULLCP, "More than one \"from\" person");
		continue;
	    }
	    from = cp;
	    rewritefrom = 1;
	    continue;
	    
	case EXTHDRSW:	/* read recipients from message */
	    extract = 1;
	    continue;
	    
	case VERBSW:	/* give blow-by-blow description */
	    verbose = 1;
	    continue;
	    
	    /* These switches have no args. */
	case QMSGSW:	/* always queue the message */
	case DAEMONSW:	/* run as a daemon & wait for SMTP */
	case INITALSW:	/* initialize the alias database */
	case QSUMSW:	/* print summary of mail queue */
	case ADRTSTSW:	/* test the addresses to debug config file */
	case CFGFRZSW:	/* create the configuration freeze file */
	case NOALISW:	/* do not do aliasing */
	case NOCONSW:	/* do not initiate immediate host connection */
	case NEWALISW:	/* run newaliases to rebuild db */
	case UFROMSW:	/* save UNIX-style From lines at front of msg*/
	case NODOTSW:	/* dots on line are not msg terminators */
	case MEOKSW:	/* ok to send to me if I'm in an alias */
	case OLDHDRSW:	/* msg may have old-style headers */
	    continue;
	    
	    /* These switches have string args. */
	case ALTALISW:	/* use alternate alias file */
	case ALTCFGSW:	/* use alternate configuration file */
	case DBGVALSW:	/* set the debug value */
	case FULLSW:	/* set full name */
	case MSGIDSW:	/* try to deliver queued msg with msg-id */
	case QTIMESW:	/* interval between queue passes */
	case DLVMODSW:	/* set the delivery mode */
	case ERRMODSW:	/* set the error mode */
	case TMPMODSW:	/* the mode to use when creating tmp files */
	case HLPFILSW:	/* the SMTP help file */
	case QDIRSW:	/* directory into which to queue messages */
	case RTMOUTSW:	/* timeout on reads */
	case SFILESW:	/* save statistics in this file */
	case MTMOUTSW:	/* timeout on messages in the queue */
	case TZSW:	/* set the name of the timezone */
	    if (*(++cp) == '\0' &&
		(!(cp = *argp++) || *cp == '-'))
		adios (NULLCP, "missing argument to %s", argp[-2]);
	    /* At this point, cp points to the argument */
	    continue;	/* Ignore */
	    
	    /* These switches have numeric args. */
	case HOPCNTSW:	/* hop count */
	case GIDSW:	/* gid when calling mailers */
	case LOGLEVSW:	/* the log level */
	case UIDSW:	/* uid when calling mailers */
	    if (*(++cp) == '\0' &&
		(!(cp = *argp++) || *cp == '-'))
		adios (NULLCP, "missing argument to %s", argp[-2]);
	    /* At this point, cp points to the numeric arg */
	    if (!isdigit(*cp))
		adios (NULLCP, "non-numeric argument to %s", argp[-2]);
	    continue;	/* Ignore */
	}
    }
    
    (void) setuid(getuid());
    
    if (verify && extract)
	adios (NULLCP, "mode not supported on header components");
    
    if (*argp == NULL && !extract)
	adios (NULLCP, "usage: /usr/lib/sendmail [flags] addr...");
    
    strcpy (msgfname, "/tmp/sendmhXXXXXX");
    if ( mktemp(msgfname) == NULL )
	adios (msgfname, "can't create msg file ");
    
    if ( (fp = fopen(msgfname,"w") ) == NULL ) {
	adios (msgfname, "error opening ");
    }
    
    doheader(argp);
    if ( verify ) {
	(void) fclose(fp);
	status = doverify();
	removemsg();
	exit ( status ) ;
    }
    dobody();
    status = sendfile();
    removemsg();
    exit ( status );
}

static void removemsg()
{
  if ( unlink(msgfname) != 0 )
      perror("unlink");
}

doheader(argp)
char **argp;
{
    char	line[BUFSIZ];
    int		gotdate, gotfrom, gotsender, gotto;
    
    /* if we're not extracting the headers from the message, then we
     * need to check to see if we need to do a "send" or a "dist".
     */
    
    if ( !extract ) {
	/* If we're doing a verify, just create a "To:" header. */
	if ( ! verify ) {
	    gotdate = gotfrom = gotto = gotsender = dodist = 0;
	    while (fgets (line, BUFSIZ, stdin) != NULL) {
		if (line[0] == '\n')		/* end of header */
		    break;
		if ( !isheader(line) )
		    break;

		/* if any of the following headers are present, then we
		 * want to do a dist.
		 */
		if ( !gotdate && uprf(line, "date") )
		    gotdate = dodist = 1;

		else if ( !gotto && (uprf(line, "to") || uprf(line, "cc")) )
		    gotto = dodist = 1;

		else if ( uprf(line, "message-id") )
		    dodist = 1;

		else if ( !gotsender && uprf(line, "sender") )
		    gotsender = dodist = 1;
		
		else if ( uprf ( line, "resent-" ) ) {
		    dodist = 1;
		    (void) fputs("Prev-", fp);
		}
		
		/* See if we are re-writing the from line */
		if ( uprf(line, "from") ) {
		    gotfrom = 1;
		    if ( rewritefrom )
			dofrom();
		    else
			(void) fputs(line,fp);
		}
		else
		    (void) fputs(line,fp);
	    }
	}
	/* Now, generate a "to" line.  The first line is easy.
	 * Write the rest of the lines with a newline/tab so that we
	 * don't accidentally write a line that's too long to be parsed
	 * by post.
	 */
	(void) fprintf (fp, "%sTo: %s", (dodist ? "Resent-" : "" ), *argp++);
	while ( *argp )
	    (void) fprintf ( fp, ",\n\t%s", *argp++ );
	(void) fputs("\n",fp);

	/* If we're doing a dist, we must have a "Date:" and "From:" field.
	 */
	if ( dodist ) {
	    if ( !gotdate )
		(void) fprintf (fp, "Date: %s\n", dtime (&lclock));
	    if ( !gotfrom )
		dofrom();
	}
#ifdef	MMDFI			/* sigh */
	if ( !gotsender )
	    (void) fprintf (fp, "Sender: %s\n", from);
#endif	MMDFI
    } else {	/* we're verifying, so just pass everything through */
	while (fgets (line, BUFSIZ, stdin) != NULL) {
	    if (line[0] == '\n')		/* end of header */
		break;
	    
	    if ( rewritefrom && uprf(line, "from"))
		dofrom();
	    else
		(void) fputs(line,fp);
	}
    }
    /* At this point, line is either a newline (end of header) or the
     * first line of the body (poorly formatted message).  If line
     * contains a line of body from a poorly formatted message, then
     * print a newline to separate the header correctly, then print
     * the body line.
     */
    if ( line[0] != '\n' )	/* i.e. a "body" line */
	(void) fputc('\n', fp);
    (void) fputs(line, fp);
}

static int isheader(s)
char *s;
{
    register char *cp;

    /* If the first character is a space, assume a continuation of a header */
    if ( isspace(*s) )
	return 1;

    /* If there's no ':', it's not a header */
    if ( (cp = index(s,':')) == NULL )
	return 0;

    /* If there's a space between BOL and ':', it's not a header */
    while ( s < cp ) {
	if ( isspace(*s) )
	    return 0;
	s++;
    }
    return 1;
}

/* This procedure does the verify and returns the status */
doverify() {
    char *command, buf[BUFSIZ], *bp;
    FILE *verfp, *popen();
    
    /* set up the command line for post */
    if ( (command = (char *)malloc((strlen(postproc) +
				    strlen(" -whom -check -verbose ") +
				    strlen(msgfname) + 1 )*sizeof(char)))
	== NULL ) {
	perror("malloc");
	return NOTOK;
    }
    
    (void) strcpy(command,postproc);
    (void) strcat(command," -whom -check ");
    if ( verbose )
	(void) strcat(command, "-verbose " );
    (void) strcat(command, msgfname);
    
    /* open up the pipe */
    if ( (verfp = popen(command,"r")) == NULL )
	return NOTOK;
    
    while ( fgets(buf, BUFSIZ, verfp) != NULL )
	/* sendmail returns:
	 *   address:  result
	 * so we need to strip the extra post headers.
	 */
	if ( verbose ) {
	    bp = buf;
	    while (isspace(*bp))
		bp++;
	    if ( *bp != '-' )
		(void) fputs(bp,stdout);
	}
    
    /* return the error status of post */
    return( pclose(verfp) >> 8 );
}

static int sendfile()
{
    char *command, buf[BUFSIZ];
    FILE *verfp, *popen();
    
    /* set up the command line for post */
    if ( (command = (char *)malloc((strlen(postproc) +
				    strlen(" -dist -verbose ") +
				    strlen(msgfname) + 1 )*sizeof(char)))
	== NULL ) {
	perror("malloc");
	return NOTOK;
    }
    
    (void) strcpy(command,postproc);
    (void) strcat(command," ");
    if ( verbose )
	(void) strcat(command, "-verbose " );
    if ( dodist )
	(void) strcat(command, "-dist " );
    (void) strcat(command, msgfname);
    
    /* open up the pipe */
    if ( (verfp = popen(command,"r")) == NULL )
	return NOTOK;
    
    while ( fgets(buf, BUFSIZ, verfp) != NULL )
	(void) fputs(buf,stdout);

    /* return the error status of post */
    return( pclose(verfp) >> 8 );
}

dofrom() {
    char	line[128];
    
    if (FullName)
	(void) sprintf(line, "From: %s <%s>\n", FullName, from);
    else
	(void) sprintf(line, "From: %s\n", from);
    (void) fputs(line, fp);
}

dobody() {
    register int    i;
    char    buffer[BUFSIZ];
    
    while (!feof (stdin) && !ferror (stdin) &&
	   (i = fread (buffer, sizeof (char), sizeof (buffer), stdin)) > 0)
	if (fwrite (buffer, sizeof (char), i , fp) != i )
	    adios (NULLCP, "Problem writing body");
    
    if (ferror (stdin))
	adios (NULLCP, "Problem reading body");
    
    if ( fclose(fp) != 0 )
	adios (NULLCP, "problem ending submission");
}

TYPESIG silentdie();

smtp()
{
    int sd,len;
    char buf[BUFSIZ], response[BUFSIZ];

    if ((sd = client(NULLCP, "tcp", "smtp", 0, response)) == NOTOK)
	adios (NULLCP, "cannot open smtp client process");

    (void) signal(SIGCHLD, silentdie);

    switch ((childid = fork())) {
	case NOTOK:
	    adios (NULLCP, "unable to fork smtp process");
	
	case OK:	/* i.e. child */
	    (void) dup2(sd,0);
	    break;

	default:	/* i.e. parent */
	    (void) dup2(sd,1);
	    break;
    }
    while ( (len = read(0, buf, BUFSIZ)) > 0)
	(void) write (1, buf, len);

    if (childid)
	(void) kill(childid, SIGHUP);

    exit(9);
}

/* ARGSUSED */
TYPESIG	die(sig)
int sig;
{
    if (fp) {
	(void) fclose(fp);
	(void) unlink(msgfname);
    }
    if (sig != SIGHUP)
	(void) fprintf(stderr, "sendmail: dying from signal %d\n", sig);
    exit(99);
}

/* ARGSUSED */

TYPESIG	silentdie(sig)
int sig;
{
    pidwait (childid, OK);
    exit(0);
}
